libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c

```diff
@@ -0,0 +1,3476 @@
+	}
+
+	if( pxDerKey != NULL )
+	{
+		vPortFree( pxDerKey );
+	}
+
+	return xResult;
+}
+
+/* Helper function for importing elliptic curve public keys from
+ * template using C_CreateObject. */
+CK_RV prvCreateECPublicKey( mbedtls_pk_context * pxMbedContext,
+							CK_ATTRIBUTE_PTR * ppxLabel,
+							CK_ATTRIBUTE_PTR *ppxClass,
+							CK_ATTRIBUTE_PTR pxTemplate,
+							CK_ULONG ulCount,
+							CK_OBJECT_HANDLE_PTR pxObject )
+{
+	CK_RV xResult = CKR_OK;
+	int lMbedReturn;
+	CK_BBOOL xBool;
+	uint32_t ulIndex;
+	CK_ATTRIBUTE xAttribute;
+	*ppxLabel = NULL;
+	*ppxClass = NULL;
+
+	/* Key will be assembled in the mbedTLS key context and then exported to DER for storage. */
+	mbedtls_ecp_keypair * pxKeyPair = ( mbedtls_ecp_keypair * ) pxMbedContext->pk_ctx;
+
+	for( ulIndex = 0; ulIndex < ulCount; ulIndex++ )
+	{
+		xAttribute = pxTemplate[ ulIndex ];
+
+		switch( xAttribute.type )
+		{
+			case ( CKA_CLASS ):
+				*ppxClass = &pxTemplate[ ulIndex ];
+			case ( CKA_KEY_TYPE ):
+
+				/* Do nothing.
+				 * Key type and class were checked previously. */
+				break;
+
+			case ( CKA_TOKEN ):
+				memcpy( &xBool, xAttribute.pValue, sizeof( CK_BBOOL ) );
+
+				if( xBool != CK_TRUE )
+				{
+					PKCS11_PRINT( ( "ERROR: Only token key creation is supported. \r\n" ) );
+					xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+				}
+
+				break;
+
+			case ( CKA_LABEL ):
+
+				if( xAttribute.ulValueLen < pkcs11configMAX_LABEL_LENGTH )
+				{
+					*ppxLabel = &pxTemplate[ ulIndex ];
+				}
+				else
+				{
+					xResult = CKR_DATA_LEN_RANGE;
+				}
+
+				break;
+
+			case ( CKA_VERIFY ):
+				memcpy( &xBool, xAttribute.pValue, xAttribute.ulValueLen );
+
+				if( xBool != CK_TRUE )
+				{
+					PKCS11_PRINT( ( "Only EC Public Keys with verify permissions supported. \r\n" ) );
+					xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+				}
+
+				break;
+
+			case ( CKA_EC_PARAMS ):
+
+				if( memcmp( ( CK_BYTE[] ) pkcs11DER_ENCODED_OID_P256, xAttribute.pValue, xAttribute.ulValueLen ) )
+				{
+					PKCS11_PRINT( ( "ERROR: Only elliptic curve P-256 is supported.\r\n" ) );
+					xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+				}
+
+				break;
+
+			case ( CKA_EC_POINT ):
+				/* The first 2 bytes are for ASN1 type/length encoding. */
+				lMbedReturn = mbedtls_ecp_point_read_binary( &pxKeyPair->grp, &pxKeyPair->Q, ( ( uint8_t * ) ( xAttribute.pValue ) + 2 ), ( xAttribute.ulValueLen - 2 ) );
+
+				if( lMbedReturn != 0 )
+				{
+					xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+				}
+
+				break;
+
+			default:
+				PKCS11_PRINT( ( "Unsupported attribute found for EC public key. %d \r\n", xAttribute.type ) );
+				xResult = CKR_ATTRIBUTE_TYPE_INVALID;
+				break;
+		}
+	}
+
+	return xResult;
+}
+
+/* Helper function for importing public keys using
+ * C_CreateObject. */
+CK_RV prvCreatePublicKey( CK_ATTRIBUTE_PTR pxTemplate,
+						  CK_ULONG ulCount,
+						  CK_OBJECT_HANDLE_PTR pxObject )
+{
+#define MAX_PUBLIC_KEY_SIZE	3000
```

Comment:
Is it the intent of this port to support 2048-bit RSA and ECDSA P256?

Reply:
yes, I will add the comment in the code in the next commit.